File: solutions/java/src/trafficsignalsystem/RedState.java
package trafficsignalsystem;

public class RedState implements SignalState {
    @Override
    public void handle(TrafficLight light, TrafficSignalController controller, Direction direction) {
        System.out.println("Direction: " + direction + " | State: RED");
        int duration = controller.getSignalDuration(direction, this);
        // After RED, move to next direction's GREEN
        Direction nextDirection = controller.getNextDirection(direction);
        controller.scheduleStateChange(
            controller.getTrafficLight(nextDirection),
            nextDirection,
            new GreenState(),
            duration
        );
    }
    @Override
    public String getName() { return "RED"; }
} 

File: solutions/typescript/src/TrafficSignalSystem/SignalState/RedState.ts
import { DirectionEnum } from "../Direction";
import TrafficLight from "../TrafficLight";
import TrafficSignalController from "../TrafficSignalController";
import GreenState from "./GreenState";
import SignalState from "./SignalState";

export default class RedState implements SignalState {
    handle(
        light: TrafficLight,
        controller: TrafficSignalController,
        direction: DirectionEnum,
    ) {
        console.log("Direction: " + direction + " | State: RED");
        const duration = controller.getSignalDuration(direction, this);

        // After RED, move to next direction's GREEN
        const nextDirection = controller.getNextDirection(direction);
        controller.scheduleStateChange(
            controller.getTrafficLight(nextDirection),
            nextDirection,
            new GreenState(),
            duration,
        );
    }

    getName() {
        return "RED";
    }
}
